force-app/main/default/classes/RD2_ChangeHistoryService.cls

```diff
@@ -70,67 +101,70 @@ public without sharing class RD2_ChangeHistoryService {
+            rdStatusReasonField => new ChangeHistoryFieldSet(RecurringDonationChangeHistory__c.NewStatusReason__c, 
+                RecurringDonationChangeHistory__c.PreviousStatusReason__c),
+            rdPaymentMethodField => new ChangeHistoryFieldSet(RecurringDonationChangeHistory__c.NewPaymentMethod__c, 
+                RecurringDonationChangeHistory__c.PreviousPaymentMethod__c),
+            rdCampaignField => new ChangeHistoryFieldSet(RecurringDonationChangeHistory__c.NewCampaign__c, 
+                RecurringDonationChangeHistory__c.PreviousCampaign__c)
+        };
+
+        public RecurringDonationChangeHistory__c setChangedFields(RecurringDonationChangeHistory__c changeHistory,
+            RD2_RecurringDonation updatedRd, RD2_RecurringDonation originalRd) {
+
+                for (SObjectField rdField : rd2ChangeHistoryFields.keySet()) {
+                    if (updatedRd.fieldChanged(rdField, originalRd)) {
+                        
+                        if (changeHistory == null) {
+                            changeHistory = initChangeHistoryRecord(updatedRd, originalRd);
+                        }
+
+                        changeHistory.put(rd2ChangeHistoryFields.get(rdField).newField, 
+                            updatedRd.getValue(rdField));
+                        changeHistory.put(rd2ChangeHistoryFields.get(rdField).previousField, 
+                            originalRd.getValue(rdField));
+                    }
                 }
 
-                rdch.put(recurringDonationToChangeHistoryFields.get(rdField).newField, 
-                    rd.getValue(rdField));
-                rdch.put(recurringDonationToChangeHistoryFields.get(rdField).previousField, 
-                    oldRd.getValue(rdField));
-            }
+                return changeHistory;
         }
 
-        if(rdch != null){
-            changeHistoryList.add(rdch);
+        public RecurringDonationChangeHistory__c initChangeHistoryRecord(RD2_RecurringDonation updatedRd, 
+            RD2_RecurringDonation originalRd){
+
+                return new RecurringDonationChangeHistory__c(
+                    RecurringDonation__c = (Id)updatedRd.getValue(npe03__Recurring_Donation__c.Id),
+                    EffectiveDate__c = (Date)updatedRd.getValue(npe03__Recurring_Donation__c.StartDate__c),
+                    NewAmount__c = (Decimal)updatedRd.getValue(rdAmountField),
+                    PreviousAmount__c = (Decimal)originalRd.getValue(rdAmountField),
+                    NewAnnualValue__c = updatedRd.getAnnualValue(),
+                    PreviousAnnualValue__c = originalRd.getAnnualValue()
+                );
         }
 
-    }
-
-    public List<RecurringDonationChangeHistory__c> getChangeHistoryRecords(){
-        return changeHistoryList;
-    }
+        public void insertChangeHistoryRecords(List<RecurringDonationChangeHistory__c> changeHistoryList){
+            insert changeHistoryList;
```

Comment:
Any thoughts on error handling here? For example, should this use allOrNone (see the UTIL_DMLService) and/or should it log insert failures if they do happen?

Reply:
Good question, error handling is something we lost by not using the TDTM wrapper to insert. Using allOrNone makes sense to me, an error here means something very unexpected happened, so we should prevent further operations.